Fix part of #21308: Full backend coverage for classroom.py and contributor_dashboard.py

Lucy-hi opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #21308.
2. This PR does the following: increase test branch coverage of core/controllers/classroom.py and core/controllers/contributor_dashboard.py to 100% by adding tests to corresponding test files

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have written tests for my code.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of 100% coverage for core/controllers/classroom.py
This is the output for running make run_tests.backend PYTHON_ARGS="--test_targets=core.controllers.classroom_test --generate_coverage_report --verbose" **before** test change

![classroom_uncovered_lines](https://github.com/user-attachments/assets/dcca3bc6-59cc-4935-8ae8-7e1cc99f9f51)
![classroom_uncovered](https://github.com/user-attachments/assets/5281674c-b736-4d46-b61a-8a21204f7384)

This is the output for running make run_tests.backend PYTHON_ARGS="--test_targets=core.controllers.classroom_test --generate_coverage_report --verbose" **after** test change
![classroom_covered](https://github.com/user-attachments/assets/6613f426-7458-49a4-b9d0-549ef903efa5)
<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of 100% coverage for core/controllers/contributor_dashboard.py
This is the output for running make run_tests.backend PYTHON_ARGS="--test_targets=core.controllers.contributor_dashboard_test --generate_coverage_report --verbose" **before** test change

![contributor_dashboard_uncovered_lines](https://github.com/user-attachments/assets/7021c66e-a4a6-4907-9946-bc0cdbd0c409)
![contributor_dashboard_uncovered](https://github.com/user-attachments/assets/9a6baef8-62cb-48b3-9b8d-09a6921b6515)

This is the output for running make run_tests.backend PYTHON_ARGS="--test_targets=core.controllers.contributor_dashboard_test --generate_coverage_report --verbose" **after** test change

![contributor_dashboard_covered](https://github.com/user-attachments/assets/4d9157d4-350a-472d-8eb6-bf50885737a8)

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->


## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Lucy-hi committed:
added tests to core/classroom_test.py, expend core/classroom.py test coverage to 100%

Lucy-hi committed:
added tests to core/contributor_dashboard_test.py, expend core/contributor_dashboard.py test coverage to 100%

Lucy-hi committed:
fix lint

Lucy-hi committed:
more fixes for linting issue

Lucy-hi committed:
fix test failure

Lucy-hi commented:
@seanlip @HardikGoyal2003  Hi, I've completed this PR. Please take a look and let me know if you have any feedback or suggestions. Thanks!

HardikGoyal2003 commented:
Hi @Lucy-hi, Thanks for the ping! A reviewer has already been assigned to this PR and will review it soon. 

@kevintab95 @StephenYu2018, PTAL!! Thanks!

StephenYu2018 commented:
@masterboy376 Can you take a first pass?

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,230 @@ def test_get(self) -> None:
         }
         self.assertDictContainsSubset(expected_dict, json_response)
 
+    def test_get_skip_no_summary_or_topic_rights(self) -> None:
```
I can't understand what this is saying. Please clarify the test name and write it as an understandable English sentence fragment. See guidance at https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests as well.

As a result, I can't tell whether the test is correct or not because I don't know what it is supposed to do.

Lucy-hi replied:
Hi, thanks for pointing that out. I intend to test get classroom data with one of the prerequisite topic has no summary. As a result, that topics should be skipped. I've changed the test name to reflect this

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,230 @@ def test_get(self) -> None:
 
+    def test_get_skip_no_summary_or_topic_rights(self) -> None:
+
+        no_summary_topic_id = topic_fetchers.get_new_topic_id()
```
I don't know what a "no summary topic" is. Please explain or choose a better name.

Lucy-hi replied:
Comments added to explain the purpose of this id.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,230 @@ def test_get(self) -> None:
+        }
+        self.assertDictContainsSubset(expected_dict, json_response)
+
+    def test_get_multiple_classrooms(self) -> None:
```
Same comment as above -- I understand what this is saying but it's not talking about which aspect of get_multiple_classrooms() you're testing. Please clarify.

Lucy-hi replied:
Thanks for pointing this out. I've changed the test name and add comments to explain the purpose of this test.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,230 @@ def test_get(self) -> None:
+                self.public_topic_id_1
+            ).to_dict()
+        )
+        public_topic_1_summary_dict = {
```
Why are you populating all these fields manually when they're basically the same as topic_summary_dict? And ditto below.

Consider Python methods like copy.deepcopy() and .extend().

Lucy-hi replied:
Got it. Thanks for the suggestion. I've updated the tests.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -381,6 +605,74 @@ def test_update_classroom_data(self) -> None:
+            'thumbnail_image', 'thumbnail_filename1', raw_thumbnail_image)
+        banner = ('banner_image', 'banner_filename1', raw_banner_image)
+        self.testapp.put(
+                    classroom_handler_url,
```
Deindent this and the next few lines so it's only 4 spaces indented from the line above. Ditto elsewhere.

Lucy-hi replied:
Thanks for pointing this out. I've updated this

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -381,6 +605,74 @@ def test_update_classroom_data(self) -> None:
 
         self.logout()
 
+    def test_update_classroom_thumbnail_data_only(self) -> None:
```
It's not clear what this is testing and how it relates to the test below. Can you use a better name to clarify?

Also I see no assertions in the test, so I'm not sure what behaviour we're trying to verify here.

Lucy-hi replied:
This test and the test below intends to update classroom data similar to test_update_classroom_data, but they only update thumbnails or the banner one at a time. There is no explicit assertion, but this test is checking if the updates will cause an exception similar to test_update_classroom_data. I've updated the test names to reflect their purpose.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -381,6 +605,74 @@ def test_update_classroom_data(self) -> None:
+
+        self.logout()
+
+    def test_update_classroom_banner_data_only(self) -> None:
```
Same issues as above.

Lucy-hi replied:
Done

seanlip reviewed (changes requested):
Reviewed classroom_test.py only; deferring to CD team for contributor_dashboard_test.py.

masterboy376 commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -321,6 +323,42 @@ def test_get_skill_opportunity_data_pagination_multiple_fetches(
             self.assertFalse(response['more'])
             self.assertIsInstance(response['next_cursor'], str)
 
+    def test_get_skill_opportunity_data_pagination_null_from_beginning(
```
The name of test function is not comprehensive enough. Try to use something like, test_get_skill_opportunity_with_pagination_set_to_zero or better.

Lucy-hi replied:
Thanks for the suggestion. It's been updated

masterboy376 commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -485,6 +523,118 @@ def test_get_reviewable_translation_opportunities_filtering_language( # pylint:
+        self.assertItemsEqual(
+            response['opportunities'], [])
+
+    def test_get_reviewable_translation_opportunities_with_null_items( # pylint: disable=line-too-long
```
test_get_reviewable_translation_opportunities_with_some_opportunities_set_to_none

Lucy-hi replied:
Thanks for the suggestion. It's been updated

masterboy376 commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -625,6 +775,26 @@ def test_pin_translation_opportunity(self) -> None:
                 csrf_token=csrf_token,
                 expected_status_int=200)
 
+    def test_pin_translation_opportunity_none_language(self) -> None:
```
test_pin_translation_opportunity_with_language_code_none

Lucy-hi replied:
Thanks for the suggestion. It's been updated

masterboy376 commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -1071,6 +1241,88 @@ def test_handler_with_exp_id_not_for_contribution_raise_exception(
+
+        self.logout()
+
+    def test_handler_translatable_item_is_data_format_list(self) -> None:
```
Again, a better name here. The name should clearly communicate the use-case which checks branches 601->599, 604->595.

Lucy-hi replied:
Thanks for the suggestion. It's been updated

masterboy376 reviewed (changes requested):
Thanks for the PR @Lucy-hi, took the first pass on contributor_dashboard_test.py. //cc @StephenYu2018 PTAL.

Lucy-hi committed:
update test contents, test names, and comments according to PR review feedbacks

Lucy-hi committed:
add more comments to explain tests

Lucy-hi committed:
fix linting issue

Lucy-hi committed:
fix python type issue

seanlip committed:
Merge branch 'develop' into backend_coverage_fix

Lucy-hi commented:
@seanlip Thanks for your feedback! I've addressed your comments and updated classroom_test.py accordingly. Please take another look when you have time.

Lucy-hi commented:
@masterboy376  Thanks for your feedback! I've addressed your comments and updated contributor_dashboard_test.py accordingly. Please take another look when you have time.

Lucy-hi commented:
> @masterboy376 Can you take a first pass?

Hi, thank you for pointing this out! I uploaded the images using GitHub's built-in functionality. It seems they are private by default, but you should be able to view them as logged in user. Could you clarify what you mean by a "public URL"? Are there recommended platforms for hosting the images? Thanks!

Lucy-hi commented:
> Hey @Lucy-hi, thanks for the PR. The images provided in the description are private. Could you please reupload the images with a public URL.

Hi @masterboy376  , thank you for pointing this out! I uploaded the images using GitHub's built-in functionality. It seems they are private by default, but you should be able to view them as logged in user. Could you clarify what you mean by a "public URL"? Are there recommended platforms for hosting the images? Thanks!

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,144 @@ def test_get(self) -> None:
+        self
+    ) -> None:
+        self.login(self.CURRICULUM_ADMIN_EMAIL, is_super_admin=True)
+        # Create a published classroom with matching url_fragment with 'math'.
```
... with a URL fragment matching 'math'.

Lucy-hi replied:
Thanks for the suggestion. It's been updated

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,144 @@ def test_get(self) -> None:
+            course_details='Course details for classroom.',
+            topic_list_intro='Topics covered for classroom'
+        )
+        # Create a unpublished classroom with unmatching url_fragment 'math'.
```
... with a URL fragment that does not match 'math'.

Lucy-hi replied:
Thanks for the suggestion. It's been updated

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,144 @@ def test_get(self) -> None:
+            topic_id_to_prerequisite_topic_ids={self.public_topic_id_2: []},
+            is_published=False
+        )
+        # Create a published classroom with unmatching url_fragment 'math'.
```
Ditto.

Lucy-hi replied:
Thanks for the suggestion. It's been updated

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +231,144 @@ def test_get(self) -> None:
+                self.public_topic_id_1
+            ).to_dict()
+        )
+        public_topic_1_summary_dict_is_published = {'is_published': True}
```
The name of this is quite confusing. I think you can just inline it.

Ditto below.

Lucy-hi replied:
Thanks for the suggestion. It's been updated

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -381,6 +519,80 @@ def test_update_classroom_data(self) -> None:
 
         self.logout()
 
+    def test_update_classroom_thumbnail_data_only_should_not_return_error(
```
Should you also be checking that the update actually worked (i.e. the data for both modified and unmodified fields are as expected at the end of the operation)?

Lucy-hi replied:
Got it. I've added checks at the end of this test and the next one to verify that the classroom data is updated and the image is uploaded correctly.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -381,6 +519,80 @@ def test_update_classroom_data(self) -> None:
+
+        self.logout()
+
+    def test_update_classroom_banner_data_only_should_not_return_error(
```
Ditto

Lucy-hi replied:
Same as above.

seanlip commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -321,6 +323,42 @@ def test_get_skill_opportunity_data_pagination_multiple_fetches(
             self.assertFalse(response['more'])
             self.assertIsInstance(response['next_cursor'], str)
 
+    def test_get_skill_opportunity_with_zero_pagination_no_opportunity_return(
```
Don't know what this name means. Please reword so it makes sense in English.

Lucy-hi replied:
Done

seanlip reviewed (changes requested):
Thanks @Lucy-hi, looks pretty good. Took another pass, PTAL.

Lucy-hi committed:
Address review feedback: update classroom_test contents and revise contributor dashboard test names.

Lucy-hi committed:
Merge remote-tracking branch 'upstream/develop' into backend_coverage_fix

masterboy376 commented on core/controllers/contributor_dashboard_test.py:
```diff
@@ -321,6 +323,42 @@ def test_get_skill_opportunity_data_pagination_multiple_fetches(
             self.assertFalse(response['more'])
             self.assertIsInstance(response['next_cursor'], str)
 
+    def test_get_skill_opportunity_with_zero_page_size_returns_empty(
```
"test_get_skill_opportunity_with_zero_page_size_returns_empty" here empty seems a  little vague. Since, response also includes other values like "more" and "next_corsor". Try "test_get_skill_opportunity_with_zero_page_size_returns_no_skill_opportunity".

masterboy376 reviewed (changes requested):
Thanks @Lucy-hi, this seems good. Just a little suggestion before I approve (your work on contributor_dashboard.py). Thanks!

Lucy-hi committed:
update test name: test_get_skill_opportunity_with_zero_page_size_returns_empty to test_get_skill_opportunity_with_zero_page_size_returns_no_opportunity

masterboy376 reviewed (approved):
Thanks! LGTM

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +232,141 @@ def test_get(self) -> None:
+        }
+        self.assertDictContainsSubset(expected_dict, json_response)
+
+    def test_get_multiple_classrooms_count_all_published_unmatching_classrooms(
```
count --> counts (since the method is a singular noun)

Lucy-hi replied:
Thanks for the suggestion! Updated.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +232,141 @@ def test_get(self) -> None:
+        self.save_new_valid_classroom(
+            classroom_id='test_id',
+            topic_id_to_prerequisite_topic_ids={
+                        self.public_topic_id_1: [],
```
Deindent so that the indentation is only 4 spaces from the line above

Lucy-hi replied:
Thanks for the suggestion! Updated.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +232,141 @@ def test_get(self) -> None:
+    ) -> None:
+        self.login(self.CURRICULUM_ADMIN_EMAIL, is_super_admin=True)
+        # Create a published classroom with a url_fragment matching 'math'.
+        self.save_new_valid_classroom(
```
You should probably specify the url_fragment explicitly here, given that it's a key part of the test.

Lucy-hi replied:
Thanks for the suggestion! Updated.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -231,6 +232,141 @@ def test_get(self) -> None:
+            **{'is_published': False}
+        )
+
+        # Should return 'test_id' class, but count all public_classrooms,
```
"class" and "classroom" are different, please standardize the terminology.

Lucy-hi replied:
Thanks for the suggestion! Updated.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -381,6 +517,118 @@ def test_update_classroom_data(self) -> None:
+        # Check new thumbnail image uploaded correctly.
+        fs = fs_services.GcsFileSystem(
+            feconf.ENTITY_TYPE_CLASSROOM, self.physics_classroom_id)
+        self.assertTrue(fs.isfile('thumbnail/%s' % 'update.svg'))
```
Here and elsewhere, you don't need the %s. Just inline the string directly into the string template. Only use the % when you are trying to include a variable.

Lucy-hi replied:
Got it. I've updated this test and the one below.

seanlip reviewed (commented):
@Lucy-hi The classroom file seems pretty good! I only have minor comments but after they're addressed, I think that file is good to go. Thanks!

Lucy-hi committed:
Update classroom_test based on review feedback

Lucy-hi committed:
Merge remote-tracking branch 'upstream/develop' into backend_coverage_fix

Lucy-hi commented:
> Thanks! LGTM

@masterboy376 Thank you for the review and approval!

Lucy-hi commented:
@seanlip PTAL, thanks!

kevintab95 reviewed (approved):
Thanks @Lucy-hi! I have the same comments as Sean in the codeowner file. Please address them before merging.

seanlip commented on core/controllers/classroom_test.py:
```diff
@@ -348,8 +350,8 @@ def test_get_multiple_classrooms_count_all_published_unmatching_classrooms(
-        # Should return 'test_id' class, but count all public_classrooms,
-        # 'science' and 'test_id'.
+        # Should return classroom with 'test_id', but count all
+        # public_classrooms('science' and 'test_id').
```
Minor nit: please add a space after "classrooms".

Lucy-hi replied:
Got it, thanks for pointing that out. I've made the update.

seanlip reviewed (approved):
Thanks @Lucy-hi! I have one minor suggestion, please fix it and we can merge. Overall it looks great, well done!

Lucy-hi committed:
reformat comment in test_get_multiple_classrooms_counts_all_published_unmatching_classrooms

Lucy-hi committed:
Merge remote-tracking branch 'upstream/develop' into backend_coverage_fix

StephenYu2018 reviewed (approved):
LGTM!

Lucy-hi commented:
@seanlip PTAL, thanks!

StephenYu2018 committed:
Merge branch 'develop' into backend_coverage_fix

seanlip commented:
Looks great. Thanks @Lucy-hi!

HardikGoyal2003 committed:
Merge branch 'develop' into backend_coverage_fix
